nemo/collections/asr/parts/submodules/tdt_loop_labels_computer.py

```diff
@@ -0,0 +1,267 @@
+        last_timesteps = out_len - 1
+
+        # masks for utterances in batch
+        active_mask: torch.Tensor = out_len > 0
+        advance_mask = torch.empty_like(active_mask)
+
+        # for storing the last state we need to know what elements became "inactive" on this step
+        active_mask_prev = torch.empty_like(active_mask)
+        became_inactive_mask = torch.empty_like(active_mask)
+
+        # loop while there are active utterances
+        first_step = True
+        while active_mask.any():
+            active_mask_prev.copy_(active_mask, non_blocking=True)
+            # stage 1: get decoder (prediction network) output
+            if first_step:
+                # start of the loop, SOS symbol is passed into prediction network, state is None
+                # we need to separate this for torch.jit
+                decoder_output, state, *_ = self.decoder.predict(
+                    labels.unsqueeze(1), None, add_sos=False, batch_size=batch_size
+                )
+                first_step = False
+            else:
+                decoder_output, state, *_ = self.decoder.predict(
+                    labels.unsqueeze(1), state, add_sos=False, batch_size=batch_size
+                )
+            decoder_output = self.joint.project_prednet(decoder_output)  # do not recalculate joint projection
+
+            # stage 2: get joint output, iteratively seeking for non-blank labels
+            # blank label in `labels` tensor means "end of hypothesis" (for this index)
+            logits = (
+                self.joint.joint_after_projection(x[batch_indices, safe_time_indices].unsqueeze(1), decoder_output,)
+                .squeeze(1)
+                .squeeze(1)
+            )
+            scores, labels = logits[:, :-duration_lengths].max(dim=-1)
+            jump_durations_indices = logits[:, -duration_lengths:].argmax(dim=-1)
+            durations = all_durations[jump_durations_indices]
+
+            # search for non-blank labels using joint, advancing time indices for blank labels
+            # checking max_symbols is not needed, since we already forced advancing time indices for such cases
+            blank_mask = labels == self._blank_index
+            # for blank labels force duration >= 1
+            durations.masked_fill_(torch.logical_and(durations == 0, blank_mask), 1)
+            time_indices_current_labels.copy_(time_indices, non_blocking=True)
+            if use_alignments:
+                if self.preserve_frame_confidence:
+                    logits = F.log_softmax(logits, dim=-1)
+                alignments.add_results_masked_(
+                    active_mask=active_mask,
+                    time_indices=time_indices_current_labels,
+                    logits=logits if self.preserve_alignments else None,
+                    labels=labels if self.preserve_alignments else None,
+                    confidence=self._get_confidence_tensor(logits) if self.preserve_frame_confidence else None,
```

Comment:
@GNroy, is this correct for TDT model? 
Should it be `self._get_confidence_tensor(F.log_softmax(logits[:, :-duration_lengths]))`? or something else?

Reply:
Let's go with `self._get_confidence_tensor(F.log_softmax(logits[:, :-duration_lengths]))` for now.

Reply:
Done, thanks!